docs/proposals/api-group.md

```diff
@@ -0,0 +1,154 @@
+
+3. meta.RESTMapper:
+
+  Each group will have its own RESTMapper (of type DefaultRESTMapper), and these mappers will be registered to pkg/api#RESTMapper (of type MultiRESTMapper).
+
+  To support identically named kinds in different groups, We need to expand the input of RESTMapper.VersionAndKindForResource from (resource string) to (group, resource string). If group is not specified and there is ambiguity (i.e., the resource exists in multiple groups), an error should be returned to force the user to specify the group.
+
+## Server-side implementation
+
+1. resource handlers' URL:
+
+  We will force the URL to be in the form of prefix/group/version/...
+
+  Prefix is used to differentiate API paths from other paths like /healthz. Groups really should use the same prefix="apis", except when backward compatibility requires otherwise. Specifically,
+
+    * for /api/v1, we set the prefix="api" (which is populated from cmd/kube-apiserver/app#APIServer.APIPrefix), group="", version="v1", so the URL remains to be /api/v1.
+
+    * for new kube API groups, we will set the prefix="apis" (we will add a field in type APIServer to hold this prefix), group=GROUP_NAME, version=VERSION. For example, the URL of the experimental resources will be /apis/experimental/v1alpha1.
+
+    * for OpenShift v1 API, because it's currently registered at /oapi/v1, to be backward compatible, OpenShift may choose prefix="osapi", group="", or prefix="", group="osapi".
+
+    * for other new third-party API, they should also use the prefix="apis" and choose the group and version. Or they can be registered through the thirdparty API plugin mechanism which is WIP in #13000.
+
+2. supporting API discovery:
+
+  Ideally, at /$path, API server should return a list of subPath so that /$path/$subPath is meaningful. Specifically,
+
+  * At /prefix/ (e.g., apis/), API server will return the supported groups.
+
+  * At /prefix/group, API server will return all supported versions of the group.
+
+  * At /prefix/group/version/, API server will return "namespaces/" and all supported non-namespaced resources in this group (see #7883). We may define a type for such a resource list in unversioned.go.
+
+  * At /prefix/group/version/namespaces/,
+
+    * if group == "admin" (assuming "namespace" belongs to group "admin"), API server will return a NameSpaceList
+
+    * if group != "admin", API server will return a list of the names of namespaces
+
+  * At /prefix/group/version/namespaces/x,
+
+    * if group == "admin", API server will return the encoded Namespace x; if additionally the query parameter apiDiscovery is set to true, i.e., at /admin/v1/namespaces/x?apiDiscovery=true, API server will return a list of all supported namespaced resources in group "admin", e.g., "limitranges".
```

Comment:
I'm not necessarily in favor of this pattern.  I think it would be more accurate to say that the root of a group (`/prefix/group/version`) the set of resources (namespaced or not) supported under that version should be listed.  Clients shouldn't have to walk into `/namespaces/` to determine that namespaced resources exist.  A client should be ideally be able to make 1 call to identify the set of available API groups and versions, and 1 call to identify the resources exposed for a particular (group,version) tuple.

I would prefer to change this section to define how `/prefix/group/version` would describe namespaced resources.  There's a lot of prior art here and I'd rather adopt something that will be familiar to clients vs inventing our own thing.

Examples:

* Google https://developers.google.com/discovery/v1/reference/apis?hl=en
* OpenStack http://specs.openstack.org/openstack/keystone-specs/api/v3/identity-api-v3.html#services-v3-services
* GitHub `curl https://api.github.com`
* The swagger result for `/swaggerapi/api/v1` - the list of resources or paths is effective - in this case, clients want to be able to find the list of resources

I think path traversal is fine for truly generic clients, but I don't think it's what a client that needs to know if "serviceAccount" is implemented should use.